mixer/adapter/solarwinds/metrics_handler.go

```diff
@@ -0,0 +1,148 @@
+// limitations under the License.
+
+package solarwinds
+
+import (
+	"context"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	"istio.io/istio/mixer/adapter/solarwinds/config"
+	"istio.io/istio/mixer/adapter/solarwinds/internal/appoptics"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/metric"
+)
+
+// MeasurementPostMaxBatchSize defines the max number of Measurements to send to the API at once
+const MeasurementPostMaxBatchSize = 1000
+
+type metricsHandlerInterface interface {
+	handleMetric(context.Context, []*metric.Instance) error
+	close() error
+}
+
+type metricsHandler struct {
+	logger     adapter.Logger
+	prepChan   chan []*appoptics.Measurement
+	stopChan   chan struct{}
+	pushChan   chan []*appoptics.Measurement
+	loopFactor *bool
+	lc         *appoptics.Client
+}
+
+func newMetricsHandler(ctx context.Context, env adapter.Env, cfg *config.Params) (metricsHandlerInterface, error) {
+	buffChanSize := runtime.NumCPU() * 10
+
+	loopFactor := true
+
+	// prepChan holds groups of Measurements to be batched
+	prepChan := make(chan []*appoptics.Measurement, buffChanSize)
+
+	// pushChan holds groups of Measurements conforming to the size constraint described
+	// by AppOptics.MeasurementPostMaxBatchSize
+	pushChan := make(chan []*appoptics.Measurement, buffChanSize)
+
+	var stopChan = make(chan struct{})
+
+	var lc *appoptics.Client
+	if strings.TrimSpace(cfg.AppopticsAccessToken) != "" {
+		lc = appoptics.NewClient(cfg.AppopticsAccessToken, env.Logger())
+
+		batchSize := cfg.AppopticsBatchSize
+		if batchSize <= 0 || batchSize > MeasurementPostMaxBatchSize {
+			batchSize = MeasurementPostMaxBatchSize
+		}
+
+		env.ScheduleDaemon(func() {
+			appoptics.BatchMeasurements(&loopFactor, prepChan, pushChan, stopChan, int(batchSize), env.Logger())
+		})
+		env.ScheduleDaemon(func() {
+			appoptics.PersistBatches(&loopFactor, lc, pushChan, stopChan, env.Logger())
+		})
+	}
+	return &metricsHandler{
+		logger:     env.Logger(),
+		prepChan:   prepChan,
+		stopChan:   stopChan,
+		pushChan:   pushChan,
+		loopFactor: &loopFactor,
+		lc:         lc,
+	}, nil
+}
+
+func (h *metricsHandler) handleMetric(_ context.Context, vals []*metric.Instance) error {
+	measurements := []*appoptics.Measurement{}
+	for _, val := range vals {
+		merticVal := h.aoVal(val.Value)
+
+		m := &appoptics.Measurement{
+			Name:  val.Name,
+			Value: merticVal,
+			Time:  time.Now().Unix(),
+			Tags:  appoptics.MeasurementTags{},
+		}
+
+		for k, v := range val.Dimensions {
+			switch vv := v.(type) {
+			case int:
+			case int32:
+			case int64:
+				m.Tags[k] = strconv.FormatInt(vv, 10)
+			case float64:
+				m.Tags[k] = strconv.FormatFloat(vv, 'f', -1, 64)
+			default:
+				m.Tags[k], _ = v.(string)
+			}
+		}
+		measurements = append(measurements, m)
+	}
+	if h.lc != nil {
+		h.prepChan <- measurements
+	}
+	return nil
+}
+
+func (h *metricsHandler) close() error {
+	close(h.prepChan)
+	close(h.pushChan)
+	close(h.stopChan)
+	*h.loopFactor = false
```

Comment:
Same comment are in the other close path, you should be synchronously waiting for the goroutines to exit before returning.